File: src/main/java/cn/ltcraft/Listener.java
package cn.ltcraft;

import cn.ltcraft.rcon.Rcon;
import java.io.IOException;
import net.mamoe.mirai.message.GroupMessage;

public class Listener {
    private Main plugin;

    public Listener(Main plugin) {
        this.plugin = plugin;
        this.registerEvents();
    }
    private void registerEvents() {
        this.plugin.getEventListener().subscribeAlways(GroupMessage.class, (event) -> {
            Rcon rcon = this.plugin.getRcon().get(event.getGroup().getId());
            if (rcon == null) {
                return;
            }
            String message = event.getMessage().contentToString();
            for (String prefix : rcon.getConfig().getStringList("prefixes")) {
                if (message.startsWith(prefix)) {
                    if (rcon.getConfig().getLongList("canPerform").contains(event.getSender().getId())) {
                        message = message.substring(prefix.length());

                        try {
                            event.getSubject().sendMessage(this.plugin.command(message, rcon));
                        } catch (IOException e) {
                            this.plugin.getLogger().info("重新连接" + rcon.getConfig().getString("serverAddr") + ":" + rcon.getConfig().getInt("serverPort") + "...");
                            try {
                                rcon.disconnect();
                            } catch (IOException ex) {

                            }
                            Rcon rcon1 = this.plugin.connected(rcon.getConfig());
                            if (rcon1 == null) {
                                event.getSubject().sendMessage("连接Rcon服务器失败！请检查密码和服务器地址连通性。");
                            } else {
                                try {
                                    this.plugin.getRcon().put(event.getGroup().getId(), rcon1);
                                    event.getSubject().sendMessage(this.plugin.command(message, rcon1));
                                } catch (IOException e2) {
                                    e2.printStackTrace();
                                    event.getSubject().sendMessage("执行失败！");
                                }
                            }
                        }
                    }
                    break;
                }
            }
        });
    }
}

File: src/main/kotlin/cn/ltcraft/rcon/Listener.kt
package cn.ltcraft.rcon

import cn.ltcraft.rcon.rcon.Rcon
import net.mamoe.mirai.event.GlobalEventChannel.subscribeAlways
import net.mamoe.mirai.event.events.MessageEvent
import java.util.regex.Pattern

/**
 * Created by Angel、 on 2022/4/17 18:18
 */
object Listener {
    fun registerEvents(){
        subscribeAlways<MessageEvent> {
            for (rcon in Main.rcons) {
                if (
                    rcon.prefixes.contains(message.contentToString().substring(0, 1)) &&
                    rcon.licensor.contains(sender.id)
                ) {
                    val rconServer = Rcon(rcon)
                    if (rconServer.isFail()){
                        subject.sendMessage("连接" + rcon.serverAddress + ":" + rcon.serverPort + "失败！请检查密码和服务器地址连通性。")
                        continue//如果玩家设置了多个服务器
                    }
                    val result = rconServer.command(message.contentToString().substring(1).trim())
                    if (result.isEmpty() || result.clean().isEmpty()){
                        subject.sendMessage("执行完成，服务器无返回！")
                    }
                    subject.sendMessage(result.clean())
                }
            }
        }
    }

    /**
     * 清理§x 颜色标签
     */
    private fun String.clean() = Pattern.compile("(?i)" + "\u00A7" + "[0-9A-FK-OR]").matcher(this).replaceAll("")

}